### ISSUE TITLE
Represent objects with both specified and unspecified members

### ISSUE THREAD

**Issue by saschanaz** [heycam/webidl#568]
I've seen specs that uses `object` to represent partially specified types.

https://www.w3.org/TR/web-animations/#the-sharedkeyframelist-interface

```webidl
// `object` here really means:
// (sequence<Keyframe> or PropertyIndexedKeyframes or SharedKeyframeList)
// as noted in https://www.w3.org/TR/web-animations/#processing-a-keyframes-argument
[Constructor (object? keyframes)]
interface SharedKeyframeList {
};
```

http://w3c.github.io/webrtc-pc/#rtcstatsreport-object

```webidl
[Exposed=Window]
interface RTCStatsReport {
    // `object` here really means `RTCStats`
    readonly maplike<DOMString, object>;
};
```

Tools e.g. [TSJS-lib-generator](https://github.com/Microsoft/TSJS-lib-generator) have to manually override those `object`s.

Can we get a better way to spec those types in IDL?

**Comment by saschanaz** [comment]
>I don't understand why they don't use the types you suggest instead

The WebRTC spec [defines RTCStatsReport as](https://w3c.github.io/webrtc-pc/#rtcstatsreport-object):

>An RTCStatsReport object is a map between strings that identify the inspected objects (id attribute in RTCStats instances), and their corresponding RTCStats-derived dictionaries.

The RTCStats-derived dictionaries are defined in [a separate spec](https://www.w3.org/TR/webrtc-stats/#stats-dictionaries) by dictionary inheritance.

```webidl
// The base dictionary in WebRTC
dictionary RTCStats {
    required DOMHighResTimeStamp timestamp;
    required RTCStatsType        type;
    required DOMString           id;
};

// The "Identifiers for WebRTC's Statistics API" spec shows several subdictionaries
dictionary RTCRtpStreamStats : RTCStats {
    unsigned long      ssrc;
    DOMString          kind;
    DOMString          transportId;
    DOMString          codecId;
    unsigned long      firCount;
    unsigned long      pliCount;
    unsigned long      nackCount;
    unsigned long      sliCount;
    unsigned long long qpSum;
};

dictionary RTCCodecStats : RTCStats: {
  // ...
};

// ...
```

So the RTCStatsReport is expected to include those dictionaries as it is, not reduced as RTCStats.

```js
const rtcStatsReport = await receiver.getStats();
rtcStatsReport.get(key).ssrc // not in base RTCStats but should exist when appropriate
```

**Comment by saschanaz** [comment]
>I don't know what's going on with the keyframes case; too much prose.

```webidl
dictionary Keyframe {
    // ... property-value pairs ...
    // i.e. DOMString propertyName
    double?             offset = null;
    DOMString           easing = "linear";
    CompositeOperation  composite;
};
```

```js
// This constructor receives `sequence<Keyframe>`, but specced as `object`
new SharedKeyframeList([
  { transform: 'translateY(0px)' }, // a key-value pair not specced in Keyframe, but has a meaning
  { transform: 'translateY(-300px)' }
]);
```

The spec's algorithm gets the ES object, run [the ES-to-dictionary conversion](https://heycam.github.io/webidl/#es-to-dictionary) explicitly (without help of IDL), and then again [extracts remaining fields](https://drafts.csswg.org/web-animations/#idl-attribute-name-to-animation-property-name) from the original object.

**Comment by Ms2ger** [comment]
Wow, who came up with that?

**Comment by bzbarsky** [comment]
Web IDL is meant to represent patterns that are:

1. Common.
2. Encouraged.

People can define arbitrarily complex processing of input, of course; that's what the `any` and `object` escape hatches are for.

In this case, we have two specs that both use `object`, but they're doing completely different processing on them.  How would you want to represent this in IDL other than as `object`?

**Comment by saschanaz** [comment]
IMO the issue there is that the ES-to-dictionary algorithm ignores any unspecified fields. If we have some way to get the unspecified fields without using `object`, that would work.

**Comment by bzbarsky** [comment]
Above we have two use cases described:

1.  The webrtc one.  This has effectively a union of dictionaries as a return value.  A union of dictionaries is not expressible in idl, so it ends up using `object`.  To produce the object, it takes whatever dictionary it has (which varies based on the map key string) and converts it to an object.  There's no "unspecified fields" involved.

2.  The keyframe case.  That case is basically trying to avoid hardcoding a list of names, for efficiency reasons.  The option of just doing a [[Get]] for every CSS property name by listing them all in the dictionary was in fact considered, and decided to be too slow.

This _could_ be addressed by something which acts like a dictionary but also snapshots the set of property names and values on the object or something.  But we have exactly one use case so far, and it's not clear that anyone else will want to do this, and if so whether they would want it to work the way the keyframe case does (e.g. ignoring prototype properties).

**Comment by saschanaz** [comment]
Adding [WebCrypto case](https://w3c.github.io/webcrypto/#cryptokey-interface) which also uses `object` for a union of dictionaries:

```webidl
[SecureContext,Exposed=(Window,Worker)]
interface CryptoKey {
  readonly attribute KeyType type;
  readonly attribute boolean extractable;
  readonly attribute object algorithm; // union of subdictionaries of KeyAlgorithm
  readonly attribute object usages; // just returns a sequence.
};
```

**Comment by bzbarsky** [comment]
Right, for the "union of dictionaries" case as a return value, `object` works ok.  The amount of spec prose you have to write in this case is the same as if you actually allowed unions of dictionaries in IDL, right?

**Comment by saschanaz** [comment]
> The amount of spec prose you have to write in this case is the same as if you actually allowed unions of dictionaries in IDL, right?

My point is for automation and tooling rather than spec authoring itself, but probably yes.

Today I saw that [Permissions API also uses `object` to receive a union of dictionaries](https://w3c.github.io/permissions/#permissions-interface):

> ```webidl
> [Exposed=(Window,Worker)]
> interface Permissions {
>   Promise<PermissionStatus> query(object permissionDesc);
> };
> ```

**Comment by Ms2ger** [comment]
Note that the union of dictionaries case is also filed in #57.
